Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+                }
+
+                for (const auto& entityComponent : entity->GetComponents())
+                {
+                    // Find if the component has a component mode using GetNumOfEventHandlers
+                    const auto entityComponentIdPair = AZ::EntityComponentIdPair(entity->GetId(), entityComponent->GetId());
+                    AddComponentButton(entityComponentIdPair);
+                }
+            }
+        }
+        else if (!newlyDeselectedEntityIds.empty())
+        {
+        /*    AZStd::vector<AZ::EntityComponentIdPair> pairIdsToRemove;
+            for (auto componentDataIt : m_addedComponents)
+            {
+                pairIdsToRemove.push_back(componentDataIt.m_pairId);
+            }
+            for (auto pairId : pairIdsToRemove)
+            {
+                RemoveComponentButton(pairId);
+            }*/
+            //v.erase(std::remove_if(v.begin(), v.end(), is_odd), v.end());
+            for (AZStd::vector<ComponentData>::reverse_iterator it = m_addedComponents.rbegin(); it != m_addedComponents.rend(); ++it)
+            {
+                RemoveComponentButton(it->m_pairId);
+            }
+        }
+    }
+
+    void ComponentModeSwitcher::AddComponentButton(const AZ::EntityComponentIdPair pairId)
+    {
+        // Check if the component has a component mode.
+        const auto numHandlers = AzToolsFramework::ComponentModeFramework::ComponentModeDelegateRequestBus::GetNumOfEventHandlers(pairId);
+        if (numHandlers < 1)
+        {
+            return;
+        }
+
+        ComponentData newComponentData = ComponentData(pairId);
+
+        // If the component has not already been added as a button, add the button
+        if (auto componentDataIt = AZStd::find_if(
+                m_addedComponents.begin(),
+                m_addedComponents.end(),
+                [newComponentData](const ComponentData& componentInfo)
+                {
+                    return componentInfo.m_component->RTTI_GetType() == newComponentData.m_component->RTTI_GetType();
+                });
+            componentDataIt == m_addedComponents.end())
+        {
+            AddSwitcherButton(newComponentData, newComponentData.m_componentName.c_str(), newComponentData.m_iconPath.c_str());
+            m_addedComponents.push_back(newComponentData);
+        }
+    }
+
+    void ComponentModeSwitcher::AddSwitcherButton(ComponentData& componentData, const char* componentName, const char* iconPath)
+    {
+        ViewportUi::ButtonId buttonId;
+        ViewportUi::ViewportUiRequestBus::EventResult(
+            buttonId,
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::CreateSwitcherButton,
+            m_switcherId,
+            iconPath,
+            componentName);
+
+        componentData.m_buttonId = buttonId;
+
+        AZStd::string buttonTooltip = AZStd::string::format(" Enter %s component mode", componentData.m_componentName.c_str());
```

Comment:
update: For UI facing things we want to use `Edit Mode` not `Component Mode` (that's the internal name unfortunately). So this would be `"Enter %s Edit Mode"` (I don't think we need the initial space either)